### PR TITLE
[EventEngine] Fix race between connection and its deadline timer on Windows

### PR DESCRIPTION
**Pull request by drfloob**

### CODE REVIEW

File: src/core/lib/event_engine/windows/windows_engine.cc [hunk 1]
```diff
@@ -43,20 +45,113 @@
...
+
+void WindowsEventEngine::ConnectionState::AbortOnConnect() {
+  on_connected_cb_.reset();
+}
+
+void WindowsEventEngine::ConnectionState::AbortDeadlineTimer() {
+  deadline_timer_cb_.reset();
+}
+
+void WindowsEventEngine::ConnectionState::OnConnectedCallback::Run() {
```
Comment (Vignesh2208) [inline src/core/lib/event_engine/windows/windows_engine.cc:117]: Nit: Probably code between WindowsEventEngine::ConnectionState::OnConnectedCallback::Run and WindowsEventEngine::ConnectionState::DeadlineTimerCallback::Run can be shared.

File: src/core/lib/event_engine/windows/windows_engine.cc [hunk 2]
```diff
@@ -253,38 +349,60 @@ bool WindowsEventEngine::IsWorkerThread() { grpc_core::Crash("unimplemented"); }
...
+    grpc_core::MutexLock lock(&connection_state->mu());
+    cancelled = CancelConnectFromDeadlineTimer(connection_state.get());
+  }
+  if (cancelled) {
+    auto cb = connection_state->TakeCallback();
+    connection_state.reset();
+    cb(absl::DeadlineExceededError("Connection timed out"));
+    return;
+  }
+  GRPC_EVENT_ENGINE_TRACE(
```
Comment (Vignesh2208) [inline src/core/lib/event_engine/windows/windows_engine.cc:400]: Nit: I dont think this needs to be traced.

**Review by Vignesh2208 — APPROVED** [review]
LGTM. Only minor nits

File: src/core/lib/event_engine/windows/windows_engine.cc (hunk 2 above)
Comment (drfloob) [reply src/core/lib/event_engine/windows/windows_engine.cc]: Trace removed.

File: src/core/lib/event_engine/windows/windows_engine.cc (hunk 1 above)
Comment (drfloob) [reply src/core/lib/event_engine/windows/windows_engine.cc:117]: To eliminate 10 SLOC duplication, options include: introducing a shared base class with protected shared state, adding friend classes/functions to access private members of the ConnectionState, or putting the engine callback invocation in a lambda. I'm not a fan of any, I think this finite (2-count) duplication is best.

**Review by drfloob — COMMENTED** [review]
Thanks for the review!

**Pull request closed without merging**